### ISSUE TITLE
Discrepency compared to https://jsonlogic.com/play.html

### ISSUE THREAD

**Issue by bradser** [json-everything/json-everything#820]
### Nuget Package

JsonLogic

### Package Version

5.3.0

### Operating System

Mac

### .Net Target (if relevant, please specify the version in the description)

.Net (5 or after)

### A clear and concise description of the bug.

Rule: { "!": [{ "var": "a" }, false] }
Data: {}

Result from https://jsonlogic.com/play.html: true
Result from https://json-everything.net/json-logic/: false

Also false via local unit test.

### What did you expect?

Results to be the same.

### Please add test code or steps to reproduce the behavior.

Put rule and data at the 2 web sites and evaluate.

### Is there any other information you'd like to share regarding this bug?

I am relatively new to JsonLogic, and evaluation of ! can be tricky I know, but regardless would expect the same results.

### Code of Conduct & Contributor Guidelines

- [X] I agree to follow this project's Code of Conduct and Contribution Guidelines.

**Comment by gregsdennis** [comment]
I think I see what's happening. They're interpreting the non-empty array as truthy.

I hate the super loose implicit casing and truthiness they have. It wreaks havoc on .net.  I regret implementing this lib 😆.

**Comment by gregsdennis** [comment]
Hm... it seems their playground (and thus their implementation) doesn't match up with what their spec says.

This actually generates an error (in my code):

> The ! rule needs an array with a single parameter.

Here's there docs: 

> Logical negation (“not”). Takes just one argument.
> Note: unary operators can also take a single, non array argument

They should be returning an error.  I'll have a look at the website to determine why it's returning `false` instead of printing an error, but I do get the error when I replicate this as a test.

**Comment by gregsdennis** [comment]
Okay, looks like my site is using the newer model-less approach.  So that's technically a bug in that.

**Comment by gregsdennis** [comment]
Oh, the JS implementation is reading an array with multiple values and then interpreting that as the value.  I think I grasped that in my first comment, then forgot it.

Why would it treat an array with a single value specially by extracting the value from it and _then_ applying truthiness?!  I still regret implementing this lib.

**Comment by gregsdennis** [comment]
But then, truthiness for an array means that it's non-empty, and the result of `[{ "var": "a" }, false]` with an empty object as data is `[null, false]`, which is non-empty, so it's truthy.  Negate that, and you get false.

I still think the JS implementation is wrong here.

**Comment by TotalTechGeek** [comment]
Hey folks, 

This is also affiliated with some quirkiness in the JSON Logic spec, discussed a bit here,
https://github.com/jwadhams/json-logic/issues/44#issuecomment-2082929035

In json-logic-js and json-logic-php,
```
{ "!": 1 }
```

is supposed to be the same as 
```
{ "!": [1] }
```

This can create an ambiguity when you use a `var` as an argument directly on the negation, if the var is an array.
```
{ "!": { var: "" } } 
```

If `{ var: "" }` is `[1]`, should it interpret it as `{ "!": [[1]] }` or as `{ "!": [1] }`? 

This is a part of the spec that's a bit ambiguous. I'm just bringing this up early because it influences the next section.

---

## THE IMPORTANT PART

Discussing **this issue** specifically, @gregsdennis I think the C# implementation might be bugged here,

To make it more clear, let me shift syntax a little. 

JSON Logic is very LISP like, converting the syntax, `{ "add": [1, 2] }` translates to `add(1, 2)` or `(add 1 2)`

`{ "not": [1] }` should be `not(1)` or `(not 1)`

If you have multiple arguments for `not`, it should ignore them, `not(1, 2)`, ONLY pays attention to the first argument. 

The mental model prescribed by JSON Logic is that the array is the set of arguments for the function; not an isolated argument.

**Comment by TotalTechGeek** [comment]
Also, I had a [similar issue myself in my implementation of JSON Logic](https://github.com/TotalTechGeek/json-logic-engine/issues/31).

**Comment by TotalTechGeek** [comment]
Also -- apologies if this comes off as topic hijacking, I can create a new issue or potentially send an email / some other form of contact,

@gregsdennis Would you be interested in some additional dev support for this module / collaborating to help address some of these ambiguities? 

I'm hoping to get a dialog going between different project maintainers of JSON Logic packages to help consolidate behavior and address some of the common issues users face, and get a compatibility table going of common extensions / behavior differences.

**Comment by gregsdennis** [comment]
> In json-logic-js and json-logic-php,
> 
>     { "!": 1 }
> 
> is supposed to be the same as
> 
>     { "!": [1] }

In that specific example, yes, they are the same.  All of the operative rules work here:

1. `1` is truthy because it's a non-zero number
1. `[1]` is truthy because it's a non-empty array
1. `[1]` decomposes to `1`

It's that last rule that is really troublesome.  With just a minor tweak, I can show you how this is inconsistent: `{ "!": 0 }` vs `{ "!": [0] }`.  Using the same rules, this breaks.

1. `0` is falsy because it's zero
1. `[0]` is truthy because it's a non-empty array
1. `[0]` decomposes to `0`

So you'd expect `{ "!": 0 }` to be equivalent to `{ "!": [0] }`, but based on the rules, it's contradictory.  _There is no solution._

The genesis of this problem is array decomposition.

---

I liked the idea of JSON Logic when I first saw it, but in implementing it, I've come to lament it.  I discovered JSON-e about a year ago, and it's a much better syntax.  It can do mostly everything JSON Logic can do, and more.

The Logic submitted in the original post `{ "!": [{ "var": "a" }, false] }` can be written in JSON-e as `{ $eval: [a, false] }`.  (`a` implicitly looks up the `a` property in the data.)  However, because `a` doesn't exist in the submitted data `{}`, an error is returned instead of just forging on to produce some value.  Using the data `{"a":1}`, you get a return of `false` because of the same truthiness (non-empty array is truthy, and then that's inverted by `!`).

---

> I'm hoping to get a dialog going between different project maintainers of JSON Logic packages to help consolidate behavior and address some of the common issues users face, and get a compatibility table going of common extensions / behavior differences.

I wish you luck.  When I implemented this library (years ago), I had reached out to the creator of JSON Logic to ask some clarifying questions, hoping to get the [docs updated](https://github.com/jwadhams/json-logic/pull/38).  It took a really long time for him to just recognize this library and [put it on the website](https://github.com/jwadhams/json-logic/pull/28).  In the end, he says the project is "stable" and he's not working on it anymore.

**Comment by TotalTechGeek** [comment]
> The genesis of this problem is array decomposition.

That's correct, and what I was trying to explain, and that's why I pointed out the `var` section that creates an ambiguity. 

`{"!": 1}` is shorthand that is *supposed to* become `{"!": [1]}` when interpreted, and the operator is supposed to invert the first argument. It's not supposed to run `![1]` unless you do `{"!": [[1]] }`

---

> I wish you luck. When I implemented this library (years ago), I had reached out to the creator of JSON Logic to ask some clarifying questions, hoping to get the https://github.com/jwadhams/json-logic/pull/38. It took a really long time for him to just recognize this library and https://github.com/jwadhams/json-logic/pull/28. In the end, he says the project is "stable" and he's not working on it anymore.

Heh -- I've been trying to reach out to him both privately and publicly for a few years myself. At this point I'm trying to go directly to other contributors and see if there's any interest in building a community & spec.

I've been trying to involve him, but it hasn't worked out. 

--- 

> I discovered JSON-e about a year ago, and it's a much better syntax. I discovered JSON-e about a year ago, and it's a much better syntax

I've personally developed an appreciation for JSON Logic and the LISP-like structure. I don't like the embedded grammar in JSON-E; though I tend to develop grammars around JSON Logic itself to make it easier to work with.
